hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/sps/TestExternalStoragePolicySatisfier.java

```diff
@@ -441,6 +442,8 @@ private void doTestWhenStoragePolicySetToCOLD() throws Exception {
 
     hdfsCluster.triggerHeartbeats();
     dfs.satisfyStoragePolicy(new Path(FILE));
+    // Assert metrics.
+    assertEquals(1, hdfsCluster.getNamesystem().getPendingSPSPaths());
     // Wait till namenode notified about the block location details
     DFSTestUtil.waitExpectedStorageType(FILE, StorageType.ARCHIVE, 3, 35000,
         dfs);
```

Comment:
Is there a race condition possible, if the storage policy gets satisfied before we get the pending sps paths? in that case the assertion shall fail I suppose?

Reply:
Thanks for your review. The method `doTestWhenStoragePolicySetToCOLD` is only called in one place. I think there should be no race condition? What do you think of this?

Reply:
> Is there a race condition possible, if the storage policy gets satisfied before we get the pending sps paths? in that case the assertion shall fail I suppose?

Thanks for your review. The method doTestWhenStoragePolicySetToCOLD is only called in one place. I think there should be no race condition? What do you think of this?

Reply:
```
    // Wait till namenode notified about the block location details
    DFSTestUtil.waitExpectedStorageType(FILE, StorageType.ARCHIVE, 3, 35000,
        dfs);
```
Here you are waiting for SPS to process the path and move the blocks to the correct place, once this is done, whether ``getPendingSPSPaths`` will still return 1? I suppose no, right? the path got processed so the count should reduce to 0.

So, my take is you don't have a control on ``    DFSTestUtil.waitExpectedStorageType(FILE, StorageType.ARCHIVE, 3, 35000,
        dfs);``, if by chance SPS process that path before your assertion then the test will fail.

I haven't gone through the code,  but that is what I felt in my initial pass, if it doesn't work this way do lemme know

Reply:
> ```
>     // Wait till namenode notified about the block location details
>     DFSTestUtil.waitExpectedStorageType(FILE, StorageType.ARCHIVE, 3, 35000,
>         dfs);
> ```
> 
> Here you are waiting for SPS to process the path and move the blocks to the correct place, once this is done, whether `getPendingSPSPaths` will still return 1? I suppose no, right? the path got processed so the count should reduce to 0.
> 
> So, my take is you don't have a control on ` DFSTestUtil.waitExpectedStorageType(FILE, StorageType.ARCHIVE, 3, 35000, dfs);`, if by chance SPS process that path before your assertion then the test will fail.
> 
> I haven't gone through the code, but that is what I felt in my initial pass, if it doesn't work this way do lemme know

Thank you @ayushtkn for your detailed explanation. I made a mistake and you are right. I'll add a new unit test to assert metrics without running SPS.